Add Scenario Editor

Jean28518 opened the pull request:
This PR comes with many changes:

- New Scenario Handling: We now have Tracks -> Scenarios -> Routes -> Selectable Time
- New Scenario Editor
- New Scenario Ingame Handling
- New Scenario File Structure
- New .trackinfo File
- New Main Menu Play Selection (Also with a proposal for a new list style)
- Station Signal Control (other PRs can be ignored)
- I started a new complete menu theme (`res://Data/UI/styles/complete_menu.theme`) (Makes it easier for huds in the future)
- Adjusted Tutorials for the new scenario file handling
- Track Editor Improvement
- Complete new Station Table Handling - (Every component except station jumping where adjusted to it, I will do this as I have time, this are just a few adjustments)
- Route baking at runtime by the Route Manager -> Old rail_route baking and handling is still active, but should be switched later to the new rail route format.
- Fixed bugs in Track Editor
- Changed the default time format in world to seconds. You are still able to convert between array and seconds mode with `Math` Libary 
- Signals now have an operation_mode: They can be a block signal, a station signal or in manual mode.

I hope I didn't forget something :)

**Attention: All Scenarios of "old" tracks need to be rebuilded.** That should be very easy and much faster than before.

---

*Edit:*
*supersedes #202*
*resolves #107*

HaSa1002 commented:
I can't say I am happy about his PR. It is huge, it does a bunch of things, we want to release next week and I have to review it. Please take my language with a little distance. I don't know how harsh it's gonna sound.

I started with checking out your branch.

> I started a new complete menu theme (res://Data/UI/styles/complete_menu.theme) (Makes it easier for huds in the future)

There have been reasons why the UI was as granualar as it has been. Take this for instance:
![grafik](https://user-images.githubusercontent.com/14185889/145486195-e2e2ee89-ef74-4740-9c02-d62ef27a4e44.png)
Compared to:
![grafik](https://user-images.githubusercontent.com/14185889/145487267-d05a388a-6b6e-42f9-ae06-14ffe763ce8c.png)

Or see this image:
![grafik](https://user-images.githubusercontent.com/14185889/145486270-83f062a2-eef3-43ed-9063-937f2fe46d44.png)

Let's stick with the screenshot for a moment.
1. The header should not be SCREAMING AT OUR USERS! Instead it should use a bigger font. No text shadows pls.
2. Why do we need the inner border? It doesn't help with readability.
3. Focus and hover style don't match
4. Focused items round corners don't fit. The only excuse I have for round corners are departure tables. That is the sole reason the top corners are rounded. I absolutley hate round corners. They waste space. Not only space but whitespace that acts better as separator.

![grafik](https://user-images.githubusercontent.com/14185889/145485429-2c091a18-ab62-44d8-a5b1-30ff195bad5c.png)

Why do we need an entirley new scenario editor separated from the current editor? Why can't they be unified with the existing editor. What is the reasoning. (Though I have to admit, I like the schematic style but not at all for this purpose). It's hard to read, too.

![grafik](https://user-images.githubusercontent.com/14185889/145485961-73c5eee3-f1bf-4bd9-86fa-e2c4f859a925.png)

Ignoring the english, but that is exactly the reason I am not happy about the split.

![grafik](https://user-images.githubusercontent.com/14185889/145486857-8dda0da7-e8ab-4518-a35a-4765bd84a823.png)

Pressing back goes back into menus that have been previously omitted. I am not sure about the spacing. But again the round corners break with the buttons.
Pressing select between every click is honestly annoying. Granted, the menu has benefits for mobile devices but why do we need to sacrifice mouse input for touch input? (I know it is a trend, but I am the one using windows :P)

> New Scenario Handling: We now have Tracks -> Scenarios -> Routes -> Selectable Time

This can be achieved way easier. I dislike the menu structure.

> New Scenario File Structure
> New .trackinfo File

Do you want to punish @DasCapschen? The idea is to move to Godot native files. Are the new files really necessary? Couldn't you have been using resources?

> Station Signal Control (other PRs can be ignored)

See, this is my problem with this PR. You should have reacted to our reviews and acted on them. They already have taken a considerable amount of time. And yet, here I am again and have to review them a third time. It's not like I am jobless.

> Complete new Station Table Handling - (Every component except station jumping where adjusted to it, I will do this as I have time, this are just a few adjustments)

Don't know what to do with the info. I guess I will know how to react once I got into code review. But fix station jumping ASAP.

> Route baking at runtime by the Route Manager -> Old rail_route baking and handling is still active, but should be switched later to the new rail route format.

Same as above. Why didn't you remove the old code completely?

> Fixed bugs in Track Editor

Namely?

> Changed the default time format in world to seconds. You are still able to convert between array and seconds mode with Math Libary

Please don't do half baked changes. Either completely or not. Why is it necessary to keep the array time stuff?

> Signals now have an operation_mode: They can be a block signal, a station signal or in manual mode.

I thought they already worked that way. But again, why in a PR named "Add Scenario Editor".

Why did you change the clear color of 2D just for the scenario editor?
![grafik](https://user-images.githubusercontent.com/14185889/145488917-75d9cfba-f6d4-4752-9b67-4b82fb7f4f6f.png)

Jean28518 commented:
The complete_menu_theme is just a proposal.
With it you are also able to overwrite some button styles e.g.
It's just to save a bit time creating huds.
I don't agree with your meaning about round corners, for example this view looks very nice in my opinion:
![grafik](https://user-images.githubusercontent.com/39700889/145596264-e86882f0-a819-4ddf-aec8-6e9a0bdd90ed.png)
But for me the UI is completely irrelevant.

> Why do we need an entirley new scenario editor separated from the current editor? Why can't they be unified with the existing editor. What is the reasoning. (Though I have to admit, I like the schematic style but not at all for this purpose). It's hard to read, too.

The track editor contains everything about track building and the scenario editor contains everything about making scenarios. With it you are able to create new scenarios for existing tracks. Without scraping the original track data. In my opinion that's very nice. Other Train Simulators have this feature too.
The scenario editor and track editor don't have the same functionality - beside the small exception with the station settings you mentioned. But that exception well thought and useful.

> Granted, the menu has benefits for mobile devices but why do we need to sacrifice mouse input for touch input? (I know it is a trend, but I am the one using windows :P)
> >New Scenario Handling: We now have Tracks -> Scenarios -> Routes -> Selectable Time

> This can be achieved way easier. I dislike the menu structure.

What's your proposal about this? What do you dislike about the current structure?

> Do you want to punish @DasCapschen? The idea is to move to Godot native files. Are the new files really necessary? Couldn't you have been using resources?
These files don't contain much path information. There is no big advantage using them. But I have no problem in changing this.

> See, this is my problem with this PR. You should have reacted to our reviews and acted on them. They already have taken a considerable amount of time. And yet, here I am again and have to review them a third time. It's not like I am jobless.

I implemented your reviews and comments.. Why isn't that PR in master? But yeah.. this PR depends on station signals... 

> Same as above. Why didn't you remove the old code completely?

Because this brings many code adjustments with it and much bug potential. For the time everything is working as usual about this topic.

> > Changed the default time format in world to seconds. You are still able to convert between array and seconds mode with Math Libary
> Please don't do half baked changes. Either completely or not. Why is it necessary to keep the array time stuff?

The array stuff is sometimes useful for example the digital clock in the displays.

> > Signals now have an operation_mode: They can be a block signal, a station signal or in manual mode.

> I thought they already worked that way. But again, why in a PR named "Add Scenario Editor".

Because the scenario manager is working with the operation mode, It's just a better declaration, how the signals are working.

> Why did you change the clear color of 2D just for the scenario editor?

SImple hack for the background of the scenario editor. For the time that's okay, we don't use that color at any other point.


> I can't say I am happy about his PR. It is huge, it does a bunch of things, we want to release next week and I have to review it. Please take my language with a little distance. I don't know how harsh it's gonna sound.

That's no excuse. Please keep kind :)

At the time I don't see a release next week.. Also without this PR there are so many things to fix and polish.
E.g. just look at the cabin. The materials are looking ugly and the cabin is not modeled. The train info (F2) has bugs.
Also the track editor at upstream is kind of unusable. E.g. there weren't any options for selecting a building or changing a value and fly around after is just pain. (I fixed many small bugs like these in this PR). 

Will start working at some adjusments in this PR tomorrow.

HaSa1002 commented on src/Editor/Editor.tscn:
```diff
@@ -12,6 +12,7 @@
 [ext_resource path="res://Editor/Modules/AddObjects.gd" type="Script" id=10]
 [ext_resource path="res://Editor/Modules/Content_Selector.tscn" type="PackedScene" id=11]
 [ext_resource path="res://Editor/Modules/BuildingSettings.tscn" type="PackedScene" id=12]
+[ext_resource path="res://Data/UI/styles/complete_menu.theme" type="Theme" id=13]
```
Buttons should be centred variants

Jean28518 replied:
done

HaSa1002 commented on src/Editor/Modules/Collider2D.gd:
```diff
@@ -0,0 +1,13 @@
+extends Area2D
+
+
+var input_handling_node
+
+
+func _on_Collider_input_event(viewport, event, shape_idx) -> void:
+	if event is InputEventMouseButton and event.button_index == BUTTON_LEFT and not event.pressed:
+		input_handling_node._item_selected(get_parent().name, get_parent().get_parent().name)
+
+func set_radius(new_radius) -> void:
+	$CollisionShape2D.shape = $CollisionShape2D.shape.duplicate(true)
+	$CollisionShape2D.shape.radius = new_radius
```
What?
Are you sure get_parent().get_parent().name is safe?

HaSa1002 replied:
and which type is input_handling_node?

Jean28518 replied:
For the use cases that is safe. 
added type to variable.

HaSa1002 commented on src/Editor/Modules/Rail2DCollider.gd:
```diff
@@ -0,0 +1,86 @@
+var input_handling_node
+
+func _ready():
+	var line_points = get_parent().points
+	var width = get_parent().width
```
That is absolutely prone to breaking.
I cannot only break but rather will break.

Jean28518 replied:
For our use cases that just works fine.

HaSa1002 commented on src/Editor/Modules/Rail2DCollider.gd:
```diff
@@ -0,0 +1,86 @@
+extends Area2D
```
Please supply a class name, format according to style guide and remove commented code.

Jean28518 replied:
done

HaSa1002 commented on src/Editor/Modules/Rail2DCollider.tscn:
```diff
@@ -0,0 +1,10 @@
+[gd_scene load_steps=2 format=2]
```
This can and should be done from within the rail collider script.

Jean28518 replied:
done, removed

HaSa1002 commented on src/Editor/Modules/RailLogic.gd:
```diff
@@ -4,6 +4,9 @@ var current_rail_logic = null
 var current_rail_logic_type = ""
 var resource_selector_called = false
 
+# We need this code for the case if you delete a signal
+func _input(event):
+	visible = is_instance_valid(current_rail_logic) and get_parent().current_tab == 2
```
No. We don't need this code. If the signal is deleted, hide should explicitly called.

Jean28518 replied:
done

HaSa1002 commented on src/Editor/Modules/RailLogic.gd:
```diff
@@ -17,7 +20,6 @@ func set_rail_logic(rail_logic):
 	$StationSettings.visible = current_rail_logic_type == RailLogicTypes.STATION
 	$SpeedLimitSettings.visible = current_rail_logic_type == RailLogicTypes.SPEED_LIMIT
 	$WarnSpeedLimitSettings.visible = current_rail_logic_type == RailLogicTypes.SPEED_LIMIT_WARNING
-	$ContactPointSettings.visible = current_rail_logic_type == RailLogicTypes.CONTACT_POINT
```
Idk this code. Why has so much been removed here?

Jean28518 replied:
Because I cleaned the Editor Functionality. So such things like signal status should be only set in the scenario editor.

HaSa1002 commented on src/Editor/Modules/scenario_editor.tscn:
```diff
@@ -0,0 +1,1458 @@
+[gd_scene load_steps=11 format=2]
```
No way I can review this file. Please break this huge scene in multiple smaller once.

Jean28518 replied:
done

HaSa1002 commented on src/Editor/Scripts/Editor.gd:
```diff
@@ -263,6 +263,8 @@ func clear_selected_object() -> void:
 	selected_object_type = ""
 	$EditorHUD.clear_current_object_name()
 
+	$EditorHUD/Dummy.grab_focus()
+
```
please remove (as well as the button). I am going to fix that in a separate PR with a built-in method.

Jean28518 replied:
done

HaSa1002 commented on src/Editor/Scripts/Editor.gd:
```diff
@@ -447,14 +446,14 @@ func add_object(complete_path: String) -> void:
 	mesh_instance.translation = position
 	$World/Buildings.add_child(mesh_instance)
 	mesh_instance.set_owner($World)
+	mesh_instance.add_child(preload("res://Data/Modules/SelectCollider.tscn").instance())
```
If you could replace that with the aabb collider, you can remove the select collider.

Jean28518 replied:
done, but we need to handle building selection for the release. don't want to do this in this PR

HaSa1002 commented on src/Editor/Scripts/Editor.gd:
```diff
@@ -447,14 +446,14 @@ func add_object(complete_path: String) -> void:
-		send_message("Can't export and test tracks in Editor. " \
-				+ "Requires subsequent PR to allow direct loading of editable " \
-				+ "maps. Ping HaSa")
+		send_message("Can't export and test tracks while running Libre TrainSim in Godot Editor. " \
+				+ "Please use a build of Libre TrainSim for exporting tracks. ")
```
That is not correct.
You can export but you can't test them.
```suggestion
		send_message("Can't test tracks if runs Libre TrainSim using the Godot Editor. " \
				+ "Please use a build of Libre TrainSim to test tracks. ")
```

HaSa1002 commented on src/Editor/Scripts/ScenarioConfiguration.gd:
```diff
@@ -0,0 +1,908 @@
+extends Panel
```
That script is 750 lines too long. I guess there is stuff in it that I don't like but YOLO.

HaSa1002 commented on src/Editor/Scripts/ScenarioMap.gd:
```diff
@@ -0,0 +1,259 @@
+
+var world
+
+onready var signal_icon = preload("res://Data/Misc/Signal_Icon.png")
+onready var signal_icon_block = preload("res://Data/Misc/Signal_Icon_block.png")
+onready var signal_icon_station = preload("res://Data/Misc/Signal_Icon_station.png")
+onready var station_image = preload("res://Data/Misc/Station_Icon.png")
+onready var station_image_selected = preload("res://Data/Misc/Station_Icon_selected.png")
+onready var contact_point_image = preload("res://Data/Misc/ContactPoint_Icon.png")
+onready var spawn_point_icon = preload("res://Data/Misc/spawn_point_icon.png")
+onready var despawn_point_icon = preload("res://Data/Misc/despawn_point_icon.png")
```
onready is unnecessary.

HaSa1002 commented on src/Editor/Scripts/scenario_map.gd:
```diff
@@ -0,0 +1,269 @@
+extends Node
```
please delete the other scenario map script.

HaSa1002 commented:
Just for the record. This review took me almost five hours.

Jean28518 commented on src/Data/Scripts/Player.gd:
```diff
@@ -777,36 +790,31 @@ func handle_signal(signal_name: String) -> void:
+			if speed == 0:
+				return
+			else:
+				Logger.warn("Route incorrect! Reached station %s but the next_station should be %s." % [signal_passed.name, world.get_signal(station_table[current_station_table_index].node_name).name], self)
```
Can't assert that, because the code line in 801 happens, when the train jumps between stations.
But will err that message.

Jean28518 commented on src/Data/Scripts/Player.gd:
```diff
@@ -1071,18 +1072,21 @@ func sort_signals_by_distance(a: Dictionary, b: Dictionary) -> bool:
 	return false
 
-
+var route: Array = [] # Array of strings. Is obsolete, like bake_route().
```
Wouldn't touch the old route baking process before the release. Now everything is working with this workaround. Would change this in v.010

Jean28518 commented on src/Data/Scripts/Player.gd:
```diff
@@ -1071,18 +1072,21 @@ func sort_signals_by_distance(a: Dictionary, b: Dictionary) -> bool:
 var baked_route_signal_positions: Dictionary = {} # dictionary of all signal positions along the route (key = signal name)
 var baked_route_is_loop: bool = false
 var complete_route_length: float = 0 # length of whole route. Exspecially used for loops in routes..
+
+# Obsolete... Should be removed in the future. This is now also calculated by the route manager. But this doesn't allow loops for the time.
+# For more information look to the comment at the variable definition of 'route_information'.
```
Wouldn't touch the old route baking process before the release. Now everything is working with this workaround. Would change this in v.010

Jean28518 commented on src/Data/Scripts/RailMap.gd:
```diff
@@ -25,8 +25,9 @@ func _ready() -> void:
 
```
I am not using the RailMap,.
It was my intentional thought, but these both two differ too much.
I kept the code inside, which could be useful, if we want to use this Map in the Track Editor e.g.

HaSa1002 replied:
The code ist not going to be useful. It's going to be dead code. If we are going to need it again then we can reimplement it. Please remove.

Jean28518 replied:
done

DasCapschen replied:
So are you using the RailMap as an Editor feature or not?
Because if you are not, then could you remove all the `if not Root.Editor` from the file and restore it to the way it was before?

Jean28518 replied:
Don't use it. already done.

Jean28518 commented on src/Data/Scripts/SifaModule.gd:
```diff
@@ -74,3 +74,8 @@ func _on_SifaTimer_timeout() -> void:
 
 	stage = 3
 	enable_emergency_brakes()
+
+
+# Used for tutorial script
+func is_sifa_requesting_user_input() -> bool:
+	return stage > 0
```
If the sifa want's to be pressed, so I can display the message in the tutorial to press the sifa key

DasCapschen replied:
Then this should be in the tutorial script, not in the Sifa Module!

Jean28518 replied:
done.

Jean28518 commented on src/Data/Scripts/Singletons/Root.gd:
```diff
@@ -106,7 +113,6 @@ func checkAndLoadTranslationsForTrain(trainDirPath: String) -> void:
 
 
 # recursion_depth = -1 -> unlimited recursion
-# found_files is return value
```
Because now the function doesn't return anything anymore..

HaSa1002 replied:
Well it does. found_files is passed as reference and it is good to mention where to find the result

Jean28518 commented on src/Data/Scripts/Singletons/Root.gd:
```diff
@@ -170,3 +176,8 @@ func set_low_resolution(value: bool) -> void:
+
+func set_current_editor_track(path_to_tscn : String):
+	current_editor_track_path = path_to_tscn.get_base_dir()
+	current_editor_track = path_to_tscn.get_file().get_basename()
```
done

Jean28518 commented on src/Data/Scripts/Singletons/content_loader.gd:
```diff
@@ -175,3 +175,42 @@ func append_content_to_global_repo(content: ModContentDefinition) -> void:
 	loaded_mods.append(content)
+
+
+func get_editor_tracks():
+	var tracks: Dictionary = {}
+	var editor_directory: String = jSaveManager.get_setting("editor_directory_path", "user://editor/")
+	var track_names: Array = []
+	var files: Array = []
+	Root.crawl_directory(files, editor_directory, ["tres"], 2)
+	for file in files:
+		if file.get_file() != "content.tres":
+			continue
+		var content = load(file) as ModContentDefinition
+		if !content:
+			continue
+		for world in content.worlds:
+			var mod_path: String = file.get_base_dir()
+			track_names.push_back(world.replace("res://Mods", \
+					mod_path.get_base_dir()).get_basename())
+			tracks[track_names.back()] = [content, mod_path]
+	return tracks
+
+
+
+
+func get_scenarios_for_track(track: String) -> Array:
+	var scenario_dir: String = track.get_base_dir().plus_file("scenarios")
+	var result: Array = []
+	Root.crawl_directory(result, scenario_dir, ["scenario"], 2)
+	return result
+
+
+
+# If train_name is part of the path, this function returns the whole path
+func find_train_path(train_name: String) -> String:
+	var train_path: String = ""
+	for available_train_path in ContentLoader.repo.trains:
+		if available_train_path.find(train_name) != -1:
+			return available_train_path
+	return ""
```
We need a place for these script functions. But we can move these functions to a new separate file if you want.
But in my opinion these functions suit to this script.

HaSa1002 replied:
I am talking about the crawling. That shouldn't happen IMO. The info you are looking for should be in the content repo.

Jean28518 replied:
That makes things much more complicated with editor.
Then a track builder can't add or remove files at the editor location easier. 
For ready game packs I can understand the sense of the the content.repo. But for editor tracks in my opnion that makes things just more complicated. 
And for us that works just fine. Why change this?

DasCapschen replied:
@HaSa1002 those look similar to functions you wrote for loading Editor maps. The content loader does not load anything from `user://editor`, so editor maps are not in the global content repo upon loading. 
see `editor_configuration.gd` function `_find_content()`.

Jean28518 replied:
Also using this function in @DasCapschen mentioned function

Jean28518 commented on src/Data/Scripts/StationJumper.gd:
```diff
@@ -10,14 +10,14 @@ signal station_index_selected(station_index)
+	current_station_indices.pop_back() # Remove endstation out of list.
 	for station_index in current_station_indices:
-		$StationJumper/ItemList.add_item(stations["stationName"][station_index])
+		$StationJumper/ItemList.add_item(player.station_table[station_index].station_name)
```
These are all dict entries... for me if I change these all to square brackets, the code gets harder to read.

Jean28518 commented on src/Data/Scripts/chunk_manager.gd:
```diff
@@ -231,13 +231,17 @@ func _finish_chunk_loading():
 # called deferred from Thread
 func _add_node_to_scene_tree(parent: String, instance: Spatial):
-	assert(world.get_node(parent).get_node_or_null(instance.name) == null)
+#	assert(world.get_node(parent).get_node_or_null(instance.name) == null)
+	if world.get_node(parent).get_node_or_null(instance.name) != null:
+		return
```
But crashing the game because of this isn't the right way. It isn't critical.

Jean28518 replied:
added `Logger.err("Tried to ad an object to world, which is already loaded. Skip loading this object...", instance)` to it.

DasCapschen replied:
Keep. The. Assert.

Jean28518 commented on src/Data/Scripts/World.gd:
```diff
@@ -297,31 +302,55 @@ func get_terrain_height_at(_position: Vector2) -> float:
+			return signal_node
+	return null
+
+# Used from scenario editor, does update assigned signals from stations
+func write_station_data(rail_logic_settings) -> void:
+	for rail_logic_name in rail_logic_settings.keys():
+		var rail_logic_node: Node = get_signal(rail_logic_name)
+		if rail_logic_node == null:
+			continue
+		if rail_logic_node.type == "Station":
+			var signal_one: Node = get_signal(rail_logic_node.assigned_signal)
+			if rail_logic_settings[rail_logic_name].overwrite:
+				rail_logic_node.assigned_signal = rail_logic_settings[rail_logic_name].assigned_signal
+				rail_logic_node.personSystem = rail_logic_settings[rail_logic_name].enable_person_system
+			var signal_two: Node = get_signal(rail_logic_node.assigned_signal)
+			if signal_one != null:
+				signal_one.update()
+			if signal_two != null:
+				signal_two.update()
```
Doesn't clash.
Save system doesn't save this at all. It's just saved in the .tscn file.

Jean28518 commented on src/Data/Scripts/World.gd:
```diff
@@ -141,101 +141,106 @@ func get_signal_scenario_data() -> Dictionary:
-	timeMinute = scenario["TimeM"]
-	timeSecond = scenario["TimeS"]
-	time = [timeHour,timeMinute,timeSecond]
+var train_spawn_information: Dictionary = {
+	time = 0,
+	player_train = false,
+	train_path = "",
+	route_name = "",
+	minimal_platform_length = 0,
+	route = [],
+	station_table = [],
+	despawn_information = {}
+}
```
because this is all important information for the train spawner in the world. It's just used of two functions in this script. Nowhere else.

Jean28518 commented on src/Data/UI/map.gd:
```diff
@@ -4,10 +4,19 @@ extends Control
+	show()
+	$FullMap.show()
+	$OverlayMap.hide()
+	map_status = MapStatus.FULL
```
Was my fault. Don't need this at all anymore.. I removed the function

Jean28518 commented on src/Data/UI/play_menu.gd:
```diff
@@ -1,122 +1,61 @@
-var currentTrain: String = ""
-var currentScenario: String = ""
-var screenshot_texture: Texture
+var selected_track: String = ""
+var selected_scenario: String = ""
+var selected_route: String = ""
+var selected_time: int = -1
+var selected_train: String = ""
```
I didn't find anything about this convention in the official style guide. So we document that convention somewhere.

Jean28518 commented on src/Data/Scripts/scenario_manager.gd:
```diff
@@ -0,0 +1,47 @@
+extends Node
```
route manager just handles the route points. Not the whole scenario or the general information about the route such as time, interval or anything else.

HaSa1002 replied:
But why doesn't do it? What is the issue with having that in a Route class?

Jean28518 replied:
You are right. Maybe we could merge it together. I will merge this both two

Jean28518 commented on src/Data/UI/main_menu.tscn:
```diff
@@ -69,6 +70,7 @@ margin_left = 21.0
 margin_top = -1.05945
 margin_right = 323.0
 margin_bottom = -1.05945
+rect_min_size = Vector2( 330, 0 )
```
Then these buttons have the same size like the new Create Button Menu. That Min size determines the width, not the height.

Jean28518 replied:
done

Jean28518 commented on src/Data/UI/main_menu.tscn:
```diff
@@ -77,42 +79,42 @@ __meta__ = {
 
 [node name="Play" type="Button" parent="Buttons"]
 margin_top = 116.0
-margin_right = 328.0
+margin_right = 330.0
```
done

Jean28518 commented on src/Data/UI/main_menu.tscn:
```diff
@@ -122,25 +124,71 @@ __meta__ = {
 text = "MENU_QUIT"
 
-[node name="About" parent="." instance=ExtResource( 9 )]
+[node name="CreateMenu" type="VBoxContainer" parent="."]
```
These are just the three main menu buttons.

Jean28518 commented on src/Data/Modules/World-Pattern.tscn:
```diff
@@ -36,4 +38,6 @@ script = ExtResource( 9 )
-[node name="jSaveModuleScenarios" parent="." instance=ExtResource( 10 )]
+[node name="RouteManager" parent="." instance=ExtResource( 4 )]
+
+[node name="ScenarioManager" parent="." instance=ExtResource( 5 )]
```
removed these nodes

Jean28518 commented on src/Data/Scripts/Singletons/Root.gd:
```diff
@@ -2,9 +2,11 @@ extends Node
+var current_scenario: String # file path
+var selected_train: String # file path
+var selected_route: String # route name
+var selected_time: int # start time of scenario in seconds
```
I don't have any plan how to get the loading screen working with scenario editor.
When we get that running, I have no problem with it.

Jean28518 commented on src/Data/Scripts/World.gd:
```diff
@@ -35,9 +32,27 @@ var personVisualInstances: Array = [
+	if get_node_or_null("RouteManager") == null:
+		add_child(preload("res://Data/Modules/RouteManager.tscn").instance())
+	if get_node_or_null("ScenarioManager") == null:
+		add_child(preload("res://Data/Modules/ScenarioManager.tscn").instance())
```
removed..

Jean28518 commented on src/Data/Scripts/World.gd:
```diff
@@ -35,9 +32,27 @@ var personVisualInstances: Array = [
+	if get_node_or_null("ScenarioManager") == null:
+		add_child(preload("res://Data/Modules/ScenarioManager.tscn").instance())
+
+	if passive:
+		return
```
yeah we could invest one hour in separting this niceley, but then we also should separate code functions between ingame and Track Editor too. 
So for me this way is okay. It doesn't waste many system resources.

Jean28518 commented on src/Data/Scripts/route_manager.gd:
```diff
@@ -0,0 +1,268 @@
+
+# (Read Only)
+func get_point(index : int) -> Dictionary:
+	if index >= get_route_size():
+		Logger.err("Point index does not exist for actual list! Aborting...", "Route Manager")
+		return {}
```
You are right, done

Jean28518 commented on src/Data/UI/play_menu.gd:
```diff
@@ -1,122 +1,61 @@
-extends PanelContainer
+extends Panel
```
What is the advantage?
With Panels I can define margins of the children very easily

HaSa1002 replied:
You only need to ensure the parent has the correct anchors. The rest is set using size flags. It often results in easier working menus.

Jean28518 replied:
![grafik](https://user-images.githubusercontent.com/39700889/146210994-ba86a536-ec77-4340-8ca0-8e480f1e9a41.png)
With Panel

![grafik](https://user-images.githubusercontent.com/39700889/146211502-f1ed1130-a730-4c29-b596-da0b73451e33.png)
With PanelContainer

Do you see the difference?
Didn't get the space with PanelContainer

Jean28518 commented on src/Data/UI/play_menu.gd:
```diff
@@ -1,122 +1,61 @@
+	update_breadcrumb()
+	$V/Tracks/H/Information.hide()
+	$V/Tracks/H/ItemList.clear()
+	if ContentLoader.repo.worlds.size() == 1:
+		selected_track = ContentLoader.repo.worlds[0]
+		$V/Tracks.hide()
+		$V/Scenarios.show()
+		update_scenarios()
```
If someone wants to export the game with only one track this would be skipped.

HaSa1002 replied:
Yes and again that won't happen. We have at least three tracks with the tutorials. In release builds we are going to have five.

Jean28518 replied:
Yes and again: If someone wants to export the game with only one track this would be skipped.

Jean28518 commented on src/Data/UI/play_menu.gd:
```diff
@@ -1,122 +1,61 @@
-	Root.checkAndLoadTranslationsForTrack(currentTrack.get_file().get_basename())
-	currentScenario = ""
-	var save_path: String = ContentLoader.repo.worlds[index].get_basename() + "-scenarios.cfg"
+func _on_Trackst_item_selected(index) -> void:
```
fixed

Jean28518 commented on src/Data/UI/play_menu.gd:
```diff
@@ -1,122 +1,61 @@
 
 
 func _make_image(path: String) -> Texture:
+	var screenshot_texture: Texture
```
Do we need this global?

Jean28518 commented on src/Data/UI/play_menu.gd:
```diff
@@ -1,122 +1,61 @@
-extends PanelContainer
+extends Panel
 
```
done

Jean28518 commented on src/Data/UI/styles/font_main_menu.tres:
```diff
@@ -0,0 +1,15 @@
+[gd_resource type="Theme" load_steps=2 format=2]
+
+[ext_resource path="res://addons/Libre_Train_Sim_Editor/Data/Misc/FontMedium.tres" type="DynamicFont" id=1]
```
done

Jean28518 commented on src/Data/Scripts/SpeedLimitSignal.gd:
```diff
@@ -20,20 +20,12 @@ func _get_type() -> String:
-		$Mesh.set_surface_material(2, $Mesh.get_surface_material(2).duplicate(true))
-		$Mesh.get_surface_material(2).albedo_texture = $Viewport.get_texture()
-		set_to_rail()
+	if Root.Editor:
+		add_child(preload("res://Data/Modules/SelectCollider.tscn").instance())
```
done

Jean28518 commented on src/Data/Scripts/SpeedLimitWarnSignal.gd:
```diff
@@ -13,19 +13,11 @@ func _get_type() -> String:
-		$Mesh.set_surface_material(2, $Mesh.get_surface_material(2).duplicate(true))
-		$Mesh.get_surface_material(2).albedo_texture = $Viewport.get_texture()
-		set_to_rail()
+	if Root.Editor:
+		add_child(preload("res://Data/Modules/SelectCollider.tscn").instance())
```
done

Jean28518 commented on src/Data/Scripts/Station.gd:
```diff
@@ -26,18 +28,14 @@ func _get_type() -> String:
-		var signals = world.get_node("Signals")
-		get_parent().remove_child(self)
-		signals.add_child(self)
+	if Root.Editor:
+		add_child(preload("res://Data/Modules/SelectCollider.tscn").instance())
```
done, with the aabb collider generator

Jean28518 commented on src/Data/Scripts/chunk_manager.gd:
```diff
@@ -231,13 +231,17 @@ func _finish_chunk_loading():
 	if instance.has_method("update"):
 		instance.update()
+	if parent == "Buildings" and Root.Editor:
+		instance.add_child(preload("res://Data/Modules/SelectCollider.tscn").instance())
```
done. Added it to line 327

Jean28518 replied:
Removed it colpletely.

DasCapschen commented on src/Data/Screens/ScreenTimeTable/screen_time_table.gd:
```diff
@@ -7,25 +7,24 @@ func _ready():
+	for i in range (current_station_table_index, station_table.size()):
 
-		if stopTypes[i] == 0 or stopTypes[i] == 2:
+		if [StopType.BEGINNING, StopType.DO_NOT_HALT].has(station_table[i].stop_type):
```
I would prefer a classic `==` check, which is most likely also faster, and more readable.

Jean28518 replied:
done. these constructs are removed.

DasCapschen commented on src/Data/Scripts/ContactPoint.gd:
```diff
@@ -48,24 +39,14 @@ func set_to_rail() -> void:
+	affectTime = d.affect_time
+	affectedSignal = d.affected_signal
+	enable_for_all_trains = d.enable_for_all_trains
+	disabled = not d.enabled
```
https://docs.godotengine.org/en/stable/getting_started/scripting/gdscript/gdscript_styleguide.html#boolean-operators

in the same vain, I would actually encourage `not` over `!`

Jean28518 commented on src/Data/Scripts/route_manager.gd:
```diff
@@ -0,0 +1,268 @@
+	approach_sound_path = "",
+#	arrival_time = 0,  # Only added by get_calculated_station_*()
+#	departure_time = 0, # Only added by get_calculated_station_*()
+}
```
![grafik](https://user-images.githubusercontent.com/39700889/146207909-c353c491-b5f9-40cc-837b-1317cdd5e251.png)

Tried it, but get errors out of this. 
Also don't know how to save it easily and load it again.
I don't have the time to invest several hours to get this running properly.

Jean28518 commented on src/Data/Scripts/route_manager.gd:
```diff
@@ -0,0 +1,268 @@
+
+
+
+var way_point_pattern: Dictionary = {
+	type = RoutePointType.WAY_POINT,
+	rail_name = "",
+}
+
+
+var spawn_point_pattern: Dictionary = {
+	type = RoutePointType.SPAWN_POINT,
+	rail_name = "",
+	distance = 0,
+	initial_speed = 0,
+	initial_speed_limit = -1,
+}
+
+var despawn_point_pattern: Dictionary = {
+	type = RoutePointType.DESPAWN_POINT,
+	rail_name = "",
+	distance = 0,
+}
```
see above

Jean28518 commented on src/Data/Scripts/select_collider.gd:
```diff
@@ -0,0 +1,14 @@
+	else:
+		mesh = get_parent().mesh
+	if mesh != null:
+		$CollisionShape.shape = mesh.create_convex_shape()
```
done

Jean28518 commented on src/Data/UI/play_menu.tscn:
```diff
@@ -1,211 +1,370 @@
+[ext_resource path="res://addons/jean28518.jTools/jSaveManager/jSaveModule.tscn" type="PackedScene" id=11]
+[ext_resource path="res://Data/Modules/RouteManager.tscn" type="PackedScene" id=12]
+
+[node name="PlayMenu" type="Panel"]
```
see discussion above

Jean28518 commented on src/Data/Modules/SelectCollider.tscn:
```diff
@@ -0,0 +1,8 @@
+[gd_scene load_steps=2 format=2]
```
done

Jean28518 commented on src/Data/Modules/Rail.tscn:
```diff
@@ -55,10 +56,11 @@ transform = Transform( 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0 )
 material_override = SubResource( 2 )
 mesh = ExtResource( 2 )
 material/0 = null
+script = ExtResource( 4 )
```
done

Jean28518 commented on src/Data/Modules/RouteManager.tscn:
```diff
@@ -0,0 +1,6 @@
+[gd_scene load_steps=2 format=2]
```
done

Jean28518 commented on src/Data/Modules/ScenarioManager.tscn:
```diff
@@ -0,0 +1,9 @@
+[gd_scene load_steps=3 format=2]
```
done

Jean28518 commented on src/Data/Scripts/World.gd:
```diff
@@ -21,7 +18,7 @@ var author: String = ""
 var description: String = ""
 
-var pendingTrains: Dictionary = {"TrainName" : [], "SpawnTime" : []}
+var pending_train_spawns: Array = []
```
done

Jean28518 commented on src/Data/Scripts/World.gd:
```diff
@@ -35,9 +32,27 @@ var personVisualInstances: Array = [
 var chunk_manager: ChunkManager = null
 
+# If the World is just use as data source (e.g. for scenario editor)
+var passive: bool = false
```
done

Jean28518 commented:
Thank you very much for your time. Sorry for sometimes being harsh in the comments.
I am already sitting about 70 hours on this PR, want to get this finally done.

Next time I will make multiple, smaller PRs. Fingers crossed.

HaSa1002 reviewed (approved):
I have no further energy for this PR. Please squash the commits.

DasCapschen reviewed (approved):
Let's get this in. We can deal with any problems in further PRs.
